### PR TITLE
Change output_declarations in function_wrapper.py to be a NamedTuple

### PR DESCRIPTION
**Pull request by zou3519**
In function_wrapper.py, I changed the output of `create_derived` to be a `NamedTuple`. This makes the types more readable.

Also did the following:
- Make the python `typing` package a build dependency. This is so that we don't have to duplicate namedtuple/NamedTuple code.
- Add mypy.ini and mypy-files.txt for global mypy options and a list of files to type check; this simplifies the mypy command. The new mypy-README.md describes how to run mypy.
- Update .travis.yml to use the simplified mypy command

cc @ezyang

### Test Plan
Built code with changes and compared the contents of `torch/lib/build/aten/src/ATen/ATen/` with a build from master (they should be identical).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -171,7 +171,7 @@ On Linux
 export CMAKE_PREFIX_PATH="$(dirname $(which conda))/../" # [anaconda root directory]
 
 # Install basic dependencies
-conda install numpy pyyaml mkl setuptools cmake cffi
+conda install numpy pyyaml mkl setuptools cmake cffi typing
```
Comment (apaszke) [inline README.md]: Is that not a standard Python module? Can we add those annotations without adding more dependencies?

Comment (apaszke) [reply README.md]: Can we at least put the import in a try catch block and print a nice error message that explicitly shows the `pip` command you can use to install the module? Something like we do to remind people to update git submodules when they're missing

Comment (zou3519) [reply README.md]: Sounds good!

**Comment by ezyang** [comment]
LGTM. @colesbury can we merge this or should it wait?

**Comment by colesbury** [comment]
Go ahead. I don't think it will conflict with Tensor/Variable merge PR

**Pull request merged**